src/clj_cbor/jump.clj

```diff
@@ -0,0 +1,98 @@
+(ns clj-cbor.jump
+  "CBOR Jump Table Decoder
+
+  Uses the first byte of an entry to determine how to decode. In many instances,
+  the first byte is sufficient to directly encode a value, but in most cases can
+  short-circuit a lot of conditional logic.
+
+  https://tools.ietf.org/html/rfc7049#appendix-B"
+  (:require
+    [clj-cbor.bytes :as bytes]
+    [clj-cbor.data.core :as data]
+    [clj-cbor.decoder :as decoder])
+  (:import java.io.DataInputStream))
+
+
+(defn uint-expr
+  "return expression that corresponds to decoding a jump entry value. Jump entries
+   either directly encode some numbers [0, 24), or specify an int-type
+   (byte, short, int, long) to read from the input. This expression will be
+   inlined into the jump entry decoding method. "
+  [int-type input-symb]
+  (let [input-symb (vary-meta input-symb assoc :tag 'java.io.DataInputStream)]
+    (cond
+      (number? int-type) int-type
+      (identical? int-type :byte)  `(.readUnsignedByte ~input-symb)
+      (identical? int-type :short) `(.readUnsignedShort ~input-symb)
+      (identical? int-type :int)   `(bit-and (.readInt ~input-symb) 0xFFFFFFFF)
+      (identical? int-type :long) `(bytes/to-unsigned-long (.readLong ~input-symb)))))
+
+
+(defmacro gen-entry
+  "generate code for jump table entry, all non-atom entries, share similar
+   structure of needing to read a uint value (with a fixed number of bytes)
+   and that value along with potentially further decoding yields value "
+  [int-type decoder-symb input-symb val-symb & result-expr]
+  (let [val-expr (uint-expr int-type input-symb)]
+    `(fn [~decoder-symb ~input-symb]
+       (let [~val-symb ~val-expr]
+         ~@result-expr))))
+
+(def int-widths [:byte :short :int :long])
+
+(defmacro gen-int-type-entries
+  "generate entry implementations for all possible initial values
+  (direct, or reading from input"
+  [start-offset decoder-symb input-symb val-symb result-expr]
+  `(list
+    ~@(map-indexed
+       (fn [idx int-type]
+         [(+ idx start-offset)
+          `(gen-entry ~int-type ~decoder-symb ~input-symb ~val-symb ~result-expr)])
+       (concat (range 0 24) int-widths))))
+
+
+(def ^:private jump-entries
+  (concat
```

Comment:
It's probably worth realizing this as a vector instead of a lazy sequence of lazy sequences, unless you want to avoid computing it until the first time it's used? Seems like that could mask errors to runtime, which is undesirable.